src/variational/advi.jl

```diff
@@ -0,0 +1,213 @@
+end
+
+ADVI() = ADVI(10, 5000)
+
+alg_str(::ADVI) = "ADVI"
+
+vi(model::Model, alg::ADVI; optimizer = ADAGrad()) = begin
+    # setup
+    var_info = VarInfo()
+    model(var_info, SampleFromUniform())
+    num_params = size(var_info.vals, 1)
+
+    dists = var_info.dists
+    ranges = var_info.ranges
+    
+    q = MeanField(zeros(num_params), zeros(num_params), dists, ranges)
+
+    # construct objective
+    elbo = ELBO()
+
+    Turing.DEBUG && @debug "Optimizing ADVI..."
+    μ, ω = optimize(elbo, alg, q, model; optimizer = optimizer)
+
+    # TODO: make mutable instead?
+    MeanField(μ, ω, dists, ranges) 
+end
+
+# TODO: implement optimize like this?
+# (advi::ADVI)(elbo::EBLO, q::MeanField, model::Model) = begin
+# end
+
+function optimize(elbo::ELBO, alg::ADVI, q::MeanField, model::Model; optimizer = ADAGrad())
+    alg_name = alg_str(alg)
+    samples_per_step = alg.samples_per_step
+    max_iters = alg.max_iters
+
+    # number of previous gradients to use to compute `s` in adaGrad
+    stepsize_num_prev = 10
+    
+    # setup
+    var_info = Turing.VarInfo()
+    model(var_info, Turing.SampleFromUniform())
+    num_params = size(var_info.vals, 1)
+
+    function f(x)
+        # extract the mean-field Gaussian params
+        μ, ω = x[1:num_params], x[num_params + 1: end]
+        
+        - elbo(q, model, μ, ω, samples_per_step)
+    end
+
+    # buffer
+    x = zeros(2 * num_params)
+
+    # HACK: re-use previous gradient `acc` if equal in value
+    # Can cause issues if two entries have idenitical values
+    vs = [v for v ∈ keys(optimizer.acc)]
+    idx = findfirst(w -> vcat(q.μ, q.ω) == w, vs)
+    if idx != nothing
+        @info "[$alg_name] Re-using previous optimizer accumulator"
+        x = vs[idx]
+    end
+    
+    diff_result = DiffResults.GradientResult(x)
+
+    # TODO: in (Blei et al, 2015) TRUNCATED ADAGrad is suggested; this is not available in Flux.Optimise
+    # Maybe consider contributed a truncated ADAGrad to Flux.Optimise
+
+    i = 0
+    prog = PROGRESS[] ? ProgressMeter.Progress(max_iters, 1, "[$alg_name] Optimizing...", 0) : 0
+
+    time_elapsed = @elapsed while (i < max_iters) # & converged # <= add criterion? A running mean maybe?
+        # TODO: separate into a `grad(...)` call; need to manually provide `diff_result` buffers
+        ForwardDiff.gradient!(diff_result, f, x)
+
+        # apply update rule
+        Δ = DiffResults.gradient(diff_result)
+        Δ = Optimise.apply!(optimizer, x, Δ)
+        @. x = x - Δ
+        
+        Turing.DEBUG && @debug "Step $i" Δ DiffResults.value(diff_result) norm(DiffResults.gradient(diff_result))
+        PROGRESS[] && (ProgressMeter.next!(prog))
+
+        i += 1
+    end
+
+    @info time_elapsed
+
+    μ, ω = x[1:num_params], x[num_params + 1: end]
+
+    return μ, ω
+end
+
+function (elbo::ELBO)(q::MeanField, model::Model, μ::Vector{T}, ω::Vector{T}, num_samples) where T <: Real
```

Comment:
I suppose the computation of the ELBO should be independent of the VI algo. If so, maybe you could move this code and use `link(rand(q))` instead of the hard-coded `center_diag_gaussian_inv`?

Reply:
I actually think this should be dependent on VI algo in many cases, and that the signature should instead be `(elbo:::ELBO)(vi::ADVI, q::MeanField, ...)`. For example, the logabsdet-jac term isn't included in general when computing the ELBO, right? What I was thinking is to have a generic implementation `(elbo::ELBO)(vi, q, ...)` in the same file as impl of `ELBO` (`objectives.jl` right now), and then the different VI algorithms can override this if necessary, e.g. `ADVI`.

BUT this is why if we get a `Transformed` type for distributions, we can actually implement the `ELBO` for general `Transformed` by adding the logabsdet-jac term and implementing `entropy(q::VariationalPosterior)` instead of hard-coding into `ELBO`.

I was under the impression that `link` is supposed to be a bijective transformation such that we get a distribution with support on the entirety of the real line; so I believe for a `Normal` it's the identity transformation, not the `center_diag_gaussian_inv`. The _reparametrization_ we do using `center_diag_guassian_inv` here is to be able to compute gradients of the parameters (`μ` and `ω`).

This is actually something I forgot to mention in the "Notes" section! There are some methods, e.g. Semi-Implicit VI, where they're not only using MC estimation of the ELBO, but instead estimating a _lower bound_ of the ELBO. It's not 100% clear to me if this should then be a separate `VariationalObjective` or we simply make `(::ELBO)(vi::SIVI, ...)` an implementation that actually computes this lower-bound of the ELBO rather than the ELBO itself.
I'm partial to a different `VariationalObjective` so as to not confuse the user.

Reply:
Yes thats a good point. Having `elbo()` dispatch on the AD sounds like a reasonable idea. You are right, `link` is a bijective, see `link` and `invlink`. This transformation is only necessary if your prior is constrained, e.g. prior of scale of a Normal. For a Normal or the location of a normal you wouldn't need a transformation as it is already unconstrained.

Regarding Semi-Implicit VI, I guess I would stick to dispatching on `elbo` and makes sure its clear in the doc-string that this function actually returns a lower bound. I think we can change this later if necessary but it is probably better to first workout a consistent API that makes sense for most VI algos.

Reply:
Great!
Hmm, so you agree that `link` and `center_diag_gaussian_inv` have different purposes?

That sounds like a good approach :+1: